### ISSUE TITLE
Summary of stocks Gains/Loss currency exchange error

### ISSUE THREAD

**Issue by h-vijayakumaran** [moneymanagerex/moneymanagerex#5833]
MMEX version:
 - [ ] 1.6.0
 - [ ] 1.6.1
 - [ ] 1.6.2
 - [ ] 1.6.3 Beta
 - [x] 1.6.4 Beta

Operating System:
 - [ ] Windows
 - [ ] Mac OSX
 - [x] Linux 

# Description of the bug

This seems to be a minor error. The (unrealized) Gain/Loss in its original currency £ (£278.96) looks fine. However, when it is listed in terms of global currency €,  the currency conversion happens incorrectly (€139.82). Note that this error exists only in the stocks summary report. In the dashboard, the conversion (£278.96 -> €314.87), which is aligned with the exchange rate.

# Reproduction

Is the bug reproducible? 
 - [x] Always 
 - [ ] Randomly 
 - [ ] Happened only once

**Comment by n-stein** [comment]
Can you share a screenshot or better yet a database that reproduces this? It looks OK to me:

![image](https://user-images.githubusercontent.com/16599929/232596700-353a9470-c1af-4575-8bf2-a9310bdf39d9.png)

**Comment by h-vijayakumaran** [comment]
The [expression](https://github.com/moneymanagerex/moneymanagerex/blob/5aa3296168d0982e8079492f30f3f667e41aaebd/src/reports/summarystocks.cpp#L77) for unrealized gain/loss 
```
m_unreal_gain_loss_sum_total += (Model_Stock::CurrentValue(stock) * today_rate - Model_Stock::InvestmentValue(stock) * purchase_rate);
```
results in the incorrect behavior, see below:
![image](https://user-images.githubusercontent.com/83648377/232635857-d15b60a4-b1fa-4d05-873b-d64801d4da50.png)
In your case the result looks fine that is probably because due to the special situation,  `today_rate == purchase_rate`. I think the correct way to go would be :
```
m_unreal_gain_loss_sum_total += (Model_Stock::CurrentValue(stock) * today_rate - Model_Stock::InvestmentValue(stock) * today_rate);
```

**Comment by n-stein** [comment]
How do you account for the unrealized forex gain/loss in that case?

**Comment by h-vijayakumaran** [comment]
I think I now understand what is going on. Then I suggest that the Stocks Gain/Loss in the dashboard is update accordingly. It seems that the forex gain/loss isn't accounted there.

**Comment by n-stein** [comment]
I think both need to be modified. The current code multiplies all held shares by the rate on the initial purchase date, but these shares could have accumulated over time with various conversion rates. It should apply the date-specific rate for each purchase/sale when calculating the invested value.

Consider the below:

1. I spend 115 EUR to purchase 100 GBP of stock (EUR/GBP rate 1.15). 
2. After some time, the stock value appreciates to 110 GBP, and the Pound has strengthened relative to the Euro so the EUR/GBP rate is now 1.20.
3. I spend an additional 120 EUR to purchase another 100 GBP of stock at the 1.20 rate.
4. After some time the stock value remains unchanged (210 GBP), but the Pound has fallen and the EUR/GBP rate is now 1.12.

If I only looked at the initial rate, my "Invested value" would be 200 GBP * 1.15 = 230 EUR, but I actually invested (100 GBP * 1.15 + 100 GBP * 1.20) = 235 EUR.

At the time of point 2 above, my current holdings are 110 GBP * 1.20 EUR/GBP = 132 EUR for an unrealized gain of 17 EUR. Without factoring in the exchange rate fluctuation, I only gained 10 GBP * 1.2 = 12 EUR, so the exchange rate increasing 500 pips in my favor contributed an additional 5 EUR to my unrealized gain.

At the time of point 4, my current holdings are 210 GBP * 1.12 EUR/GBP = 235.2 EUR. Since my invested value was 235 EUR, the forex rate change has actually erased almost all of the 10 GBP gain in the stock value.

If I only looked at the initial 1.15 rate to calculate my invested value it seems I came out 5 EUR ahead, but in reality I basically broke even.

**Comment by n-stein** [comment]
Similarly the Realized Gain/Loss should take the date of the transactions into account. If I realize 10 GBP in gains at a rate of 1.2 my Realized Gain/Loss is 12 EUR. If the rate then drops to 1.12 I still realized 12 EUR in gains, not 11.2.

**Comment by n-stein** [comment]
But perhaps this isn't the right way to present it at all. It is confusing to see a report that looks like this:

![image](https://user-images.githubusercontent.com/16599929/232871465-a4060af9-b32f-4851-97b5-9a209c6c4b27.png)

In this case there is no account currency realized gain/loss because I bought & sold at the same GBP price, but there IS a realized gain in base currency because of a 500 pip rate increase between buying and selling. Perhaps this should be indicated in a separate column, or noted explicitly in the totals:

![image](https://user-images.githubusercontent.com/16599929/232884145-284b64f3-1753-463b-9edd-5e6b70020f36.png)

**Comment by h-vijayakumaran** [comment]
I hadn't thought about the accumulation aspect, but it makes complete sense. The above modifications looks fine to me, and it conveys the matter clearly. 
BTW, is the realized gain in the example a numerical error?

**Comment by n-stein** [comment]
> is the realized gain in the example a numerical error?

No, I added a sell transaction to get a screenshot where the account gain was 0 but there was a forex gain:

1. Purchase 10 shares @ 10 GBP/share with EUR/GBP = 1.15
2. Sell 1 share @ 10 GBP/share with EUR/GBP = 1.2
3. Purchase 9.0909 shares @ 11 GBP/share with EUR/GBP = 1.2

Since I bought & sold for 10 GBP/share I had no realized gain, but since the FX rate increased from 1.15 to 1.2 the 10 GBP I sold was worth 0.5 EUR more than the 10 GBP I purchased.

However, there IS an error in the screenshot. The increase in share price from 10 to 11 GBP should have given me a 9 GBP unrealized gain from my remaining 9 shares of the initial purchase, but you can see that the unrealized gain is showing as 9.48 GBP. This is because the "Initial Value" is being calculated incorrectly. It should be 190 GBP but is reporting 189.52. This is because I entered transaction 2 after transaction 3 and the system is not sorting them by date first when determining the initial value. My avg share price at the time of the sale should be 10 GBP, but it is incorrectly including the purchase in point 3 so the average is becoming 200/19.0909 = 10.48 instead. I will log a bug for this. Here is the corrected screenshot:

![image](https://user-images.githubusercontent.com/16599929/232903861-8a61a4b3-968a-4cba-b5f8-21d244b929c9.png)

**Comment by n-stein** [comment]
Individual daily FX rates are now considered instead of broadly using the FX rate of the initial purchase to calculate gain/loss.

**Comment by h-vijayakumaran** [comment]
Looks great! I will check it out when the linux build is ready
